File: producer/js/src/ast.ts
import {ANTLRInputStream, CommonToken, CommonTokenStream} from 'antlr4ts';
import {ComposeLexer} from "./ts/ComposeLexer";
import {
    CompilationUnitContext,
    ComposeParser,
    EntityCallContext,
    EntityModifierContext,
    ParameterListContext
} from "./ts/ComposeParser";
import {ParseTreeWalker} from "antlr4ts/tree";
import {ComposeParserListener} from "./ts/ComposeParserListener";
import {Vocabulary} from "antlr4ts/Vocabulary";

interface IEntityFlow {
    Entity: String
    Calls: ICodeCall[]
}

class EntityFlow implements IEntityFlow {
    Calls: ICodeCall[] = [];
    Entity: String;
}

interface ICodeCall {
    FunctionName: string
    Parameters: IParamType[]
}

class CodeCall implements ICodeCall {
    FunctionName: string;
    Parameters: IParamType[];
}

interface IParamType {
    TypeName: string
    TypeValue: string
}

class ParamType implements IParamType {
    TypeName: string;
    TypeValue: string;
}

class ComposeListener implements ComposeParserListener {
    public entity: EntityFlow;
    private vocabulary: Vocabulary;

    constructor(vocabulary: Vocabulary) {
        this.vocabulary = vocabulary;
        this.entity = new EntityFlow();
    }

    enterEntityModifier(ctx: EntityModifierContext): void {
        this.entity.Entity = ctx.text;
    }

    enterEntityCall(ctx: EntityCallContext): void {
        let codeCall = new CodeCall();
        codeCall.FunctionName = ctx.identifier().text;
        codeCall.Parameters = [];

        for (let literal of (ctx.parameterList() as ParameterListContext).literal()) {
            let tokenIndex = (literal.children[0].payload as CommonToken).type;
            let symbolicName = this.vocabulary.getSymbolicName(tokenIndex);

            let typeName = "";
            switch (symbolicName) {
                case "IntegerLiteral":
                    typeName = "int"
                    break;
                case "FloatingPointLiteral":
                    typeName = "float";
                    break;
                case "BooleanLiteral":
                    typeName = "bool";
                    break;
                case "CharacterLiteral":
                    typeName = "char";
                    break;
                case "StringLiteral":
                    typeName = "string";
                    break;
                case "NullLiteral":
                    typeName = "null";
                    break;
                default:

            }

            let paramType = new ParamType();
            paramType.TypeName = typeName;
            paramType.TypeValue = literal.text;
            codeCall.Parameters.push(paramType)
        }

        this.entity.Calls.push(codeCall)
    }
}

export function Compile(str: string) {
    let inputStream = new ANTLRInputStream(str);
    let lexer = new ComposeLexer(inputStream);
    let tokenStream = new CommonTokenStream(lexer);
    let parser = new ComposeParser(tokenStream);
    let tree = parser.compilationUnit();

    const listener = new ComposeListener(parser.vocabulary);
    ParseTreeWalker.DEFAULT.walk(listener as any, tree)

    return listener.entity
}

File: producer/go/ast.go
package producer

import (
	parser "dsl-compose/libs/go"
	"github.com/antlr/antlr4/runtime/Go/antlr"
)

type EntityFlow struct {
	Entity string
	Calls  []CodeCall
}

type CodeCall struct {
	FunctionName string
	Parameters   []ParamType
}

type ParamType struct {
	TypeName  string
	TypeValue string
}

type EntityListener struct {
	*parser.BaseComposeParserListener
	flow          EntityFlow
	ruleNames     []string
	symbolicNames []string
}

func NewTreeShapeListener(ruleNames []string, symbolicNames []string) *EntityListener {
	return &EntityListener{
		ruleNames:     ruleNames,
		symbolicNames: symbolicNames,
		flow:          EntityFlow{},
	}
}

func (t *EntityListener) EnterEntityModifier(ctx *parser.EntityModifierContext) {
	t.flow.Entity = ctx.GetText()
}

func (t *EntityListener) EnterEntityCall(ctx *parser.EntityCallContext) {
	call := CodeCall{
		FunctionName: ctx.Identifier().GetText(),
	}

	paramList := ctx.ParameterList().(*parser.ParameterListContext)
	for _, lit := range paramList.AllLiteral() {
		literal := lit.(*parser.LiteralContext)

		payload := literal.GetChildren()[0].GetPayload()
		index := payload.(*antlr.CommonToken).GetTokenType()

		typeName := t.ruleNames[literal.GetRuleIndex()]

		switch t.symbolicNames[index] {
		case "IntegerLiteral":
			typeName = "int"
		case "FloatingPointLiteral":
			typeName = "float"
		case "BooleanLiteral":
			typeName = "bool"
		case "CharacterLiteral":
			typeName = "char"
		case "StringLiteral":
			typeName = "string"
		case "NullLiteral":
			typeName = "null"
		default:

		}

		paramType := ParamType{
			TypeName:  typeName,
			TypeValue: literal.GetText(),
		}
		call.Parameters = append(call.Parameters, paramType)
	}

	t.flow.Calls = append(t.flow.Calls, call)
}

func (t *EntityListener) Entity() EntityFlow {
	return t.flow
}

func Compile(str string) EntityFlow {
	input := antlr.NewInputStream(str)
	lexer := parser.NewComposeLexer(input)
	stream := antlr.NewCommonTokenStream(lexer, 0)

	p := parser.NewComposeParser(stream)
	p.AddErrorListener(antlr.NewDiagnosticErrorListener(true))

	p.BuildParseTrees = true
	tree := p.CompilationUnit()

	listener := NewTreeShapeListener(p.GetRuleNames(), p.GetSymbolicNames())
	antlr.ParseTreeWalkerDefault.Walk(listener, tree)

	return listener.Entity()
}
